bundles/binding/org.openhab.binding.enphaseenergy/src/main/java/org/openhab/binding/enphaseenergy/internal/EnphaseenergyBinding.java

```diff
@@ -0,0 +1,185 @@
+    protected static final String CONFIG_KEY = "key";
+
+	private String	user_id;
+	private String 	key;
+
+    /**
+     * The refresh interval which is used to poll values from the Enphase Energy server
+     * (optional, defaults to 900000ms)
+     */
+    private long refreshInterval = 900000;
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    protected String getName() {
+        return "Enphaseenergy Refresh Service";
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    protected long getRefreshInterval() {
+        return this.refreshInterval;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @SuppressWarnings("incomplete-switch")
+    @Override
+    protected void execute() {
+        logger.debug("Querying Enphase Energy API");
+		try {
+			// Find all systemId for active items 
+			Set<Integer>	systemIds = new HashSet<Integer>();
+			for (final EnphaseenergyBindingProvider provider : this.providers) {
+				for (final String itemName : provider.getItemNames()) {
+					final Integer systemId = provider.getSystemId(itemName);
+					systemIds.add(systemId);
+				}
+			}
+			for (final Integer systemId : systemIds) {
+				final SystemsRequest request = new SystemsRequest(this.key, this.user_id, systemId);
+				logger.debug("Request: {} as {}", request, request.prepare());
+
+				final SystemsResponse response = request.execute();
+				logger.debug("Response: {}", response);
+
+				for (final EnphaseenergyBindingProvider provider : this.providers) {
+					for (final String itemName : provider.getItemNames()) {
+						final Integer	itemSystemId = provider.getSystemId(itemName);
+						if(!systemId.equals(itemSystemId)) {
+							continue;
+						}
+						State state = null;
+						final EnphaseenergyItemType itemType = provider.getItemType(itemName);
+						logger.debug("itemName {} for {} and {}", itemName, systemId, itemType);
+						switch (itemType) {
+						case MODULES:
+							state = new DecimalType(response.getModules());
+							break;
+						case SIZE_W:
+							state = new DecimalType(response.getSize_w());
+							break;
+						case CURRENT_POWER:
+							state = new DecimalType(response.getCurrent_power());
+							break;
+						case ENERGY_TODAY:
+							state = new DecimalType(response.getEnergy_today() * 0.001);
+							break;
+						case ENERGY_LIFETIME:
+							state = new DecimalType(response.getEnergy_lifetime() * 0.001 * 0.001);
+							break;
+						case SUMMARY_DATE:
+							state = new DateTimeType(response.getSummary_date());
+							break;
+						case SOURCE:
+							state = new StringType(response.getSource());
+							break;
+						case STATUS:
+							state = new StringType(response.getStatus());
+							break;
+						case OPERATIONAL_AT:
+							state = new DateTimeType(response.getOperational_at());
+							break;
+						case LAST_REPORT_AT:
+							state = new DateTimeType(response.getLast_report_at());
+							break;
+						}
```

Comment:
for the sake of readability, extract this switch-case into its own method like createStateFromType() (or alike)